security-csrf/src/main/java/io/micronaut/security/csrf/resolver/FieldCsrfTokenResolver.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.security.csrf.resolver;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.async.publisher.Publishers;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.ServerHttpRequest;
+import io.micronaut.http.body.ByteBody;
+import io.micronaut.http.body.CloseableByteBody;
+import io.micronaut.security.csrf.CsrfConfiguration;
+import jakarta.inject.Singleton;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+
+/**
+ * Resolves a CSRF token from a form-urlencoded body using the {@link ServerHttpRequest#byteBody()} API.
+ *
+ * @since 2.0.0
+ */
+@Requires(classes = HttpRequest.class)
+@Requires(property = "micronaut.security.csrf.token-resolvers.field.enabled", value = StringUtils.TRUE, defaultValue = StringUtils.TRUE)
+@Singleton
+class FieldCsrfTokenResolver implements ReactiveCsrfTokenResolver<HttpRequest<?>> {
+    private final CsrfConfiguration csrfConfiguration;
+
+    FieldCsrfTokenResolver(CsrfConfiguration csrfConfiguration) {
+        this.csrfConfiguration = csrfConfiguration;
+    }
+
+    @Override
+    @Singleton
+    public Publisher<String> resolveToken(HttpRequest<?> request) {
+        if (request instanceof ServerHttpRequest<?> serverHttpRequest) {
+            return resolveToken(serverHttpRequest);
+        }
+        return Publishers.empty();
+    }
+
+    private Publisher<String> resolveToken(ServerHttpRequest<?> request) {
```

Comment:
@yawkat is this the idiomatic way to get a form-url-encoded body field ?

Reply:
can't use toByteArrayPublisher because that may split form fields in the middle. have to buffer all of it. i changed it.

i wouldn't call this the idiomatic way, there simply is no idiomatic way to access form data in a server-independent way yet. but i have no better idea for now than this.

Reply:
I think we should add a way to resolve the form data to map in a server-independent way for netty I guess it can be done by splitting the body and buffering.

Not sure how to handle multiple part requests however.

Reply:
I am going to try to create a PR in core. Currently, `ServerHttpRequest` is netty only.

Reply:
see: https://github.com/micronaut-projects/micronaut-core/pull/11272

Reply:
@sdelamo also, ServerHttpRequest is not netty only. servlet also implements it, and so do the cloud modules

Reply:
I have updated the code to use the new [FilterBodyParser API](https://github.com/micronaut-projects/micronaut-security/pull/1826/commits/94beaf6f0061ab55ba53bfcab13c8f99c47fe207) and changed the [reactive API to use CompletableFuture](https://github.com/micronaut-projects/micronaut-security/pull/1826/commits/7f708037a5abd490a057f2c3b49fca63ddbe7503).